torch/_expanded_weights/expanded_weights_impl.py

```diff
@@ -0,0 +1,203 @@
+import torch
+import functools
+import warnings
+
+from typing import Callable, Dict, List, Union
+
+HANDLED_FUNCTIONS: Dict[Callable, torch.autograd.Function] = {}
+
+def implements_per_sample_grads(torch_function):
+    @functools.wraps(torch_function)
+    def decorator(autograd_func):
+        HANDLED_FUNCTIONS[torch_function] = autograd_func
+        return autograd_func
+    return decorator
+
+# ExpandedWeight represents a weight (parameter) Tensor that has an expanded
+# batch dimension. Operations on the ExpandedWeight Tensor take advantage of
+# how the batch dimension is expanded by de-expanding the weight before
+# computation. A subsequent call to .backward() computes gradients for
+# ExpandedWeight. Those gradients are equivalent to per-sample-grads for the
+# unexpanded weight Tensors.
+#
+# ExpandedWeight has a fallback that does the forward + backward computation.
+# The backward computation is not optimized: it runs torch.autograd.grad in
+# a loop. To optimize the backward computation further, we must register
+# overrides for specific operators.
+#
+# This is a __torch_function__ object but it could have also been a Tensor Extension
+# with a dispatch key.
+class ExpandedWeight(torch.Tensor):
+    handled_functions = HANDLED_FUNCTIONS
+
+    # needed for conv2d default kwargs
+    conv_kwarg_options = ['stride', 'padding', 'dilation', 'groups']
+    conv_kwarg_defaults = {'stride': 1, 'padding': 0, 'dilation': 1, 'groups': 1}
+
+    def __new__(cls, orig_weight, batch_size):
+        ret = torch.Tensor._make_subclass(cls, orig_weight.detach(), orig_weight.requires_grad)
+        if not isinstance(orig_weight, torch.Tensor):
+            raise RuntimeError(f"Can only make ExpandedWeights of Tensors, got {type(orig_weight).__name__}")
+        cls.batch_size = batch_size
+        cls.orig_weight = orig_weight
+        return ret
+
+    @classmethod
+    def __torch_function__(cls, func, _, args=(), kwargs=None):
+        if kwargs is None:
+            kwargs = {}
+        if func not in cls.handled_functions:
+            warnings.warn(f"don't have custom implementation for function {func.__name__}. Using slow fallback")
+            return SlowFallback.apply(func, *(args + tuple(kwargs.values())))
+        if func == torch.nn.functional.conv2d:
+            remaining_kwargs = 7 - len(args)
+            remaining_kwargs_options = cls.conv_kwarg_options[4 - remaining_kwargs:]
+            kwargs = {key: cls.conv_kwarg_defaults[key] for key in remaining_kwargs_options} | kwargs
```

Comment:
Hmm, do we need to do this default arg thing for all functions that have default arguments?

Reply:
So this was confusing to say the least because other functions have default arguments (and have tests that don't pass all of those default arguments) but none of them were running into the problem that conv2d was. I'm not sure if this will be fixed by the convolution refactor. My best guess was that the `handle_torch_function` call was not passing some default arguments because [others are always passing them](https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L2375-L2387) but it was a little hard to verify this in the generated code.

Reply:
Seems like a general problem: https://github.com/pytorch/pytorch/issues/45565

Reply:
If there are not too many of these (e.g. just conv2d) -- we could use the hack that you're doing here. We should add some parity testing though just in case the defaults ever change.

iirc torch.fx had a similar problem and they wrote something like their own "default argument filler" - maybe Horace or James would know more but they're out at the moment.

Reply:
Resolved offline: OpInfos have the case with defaults if we trust them.